### PR TITLE
Jay - PERMISSION: Add Time Entry (Others) 

### PR DESCRIPTION
**Pull request by Alforoan**
# Description
This PR enables the user with the Add Time Entry (Others) permission to be able to add a time entry form on any user via their dashboard.

## Related PRS (if any):
This frontend PR is related to the [#727](https://github.com/OneCommunityGlobal/HGNRest/pull/727) backend PR.

…

## Main changes explained:
- Added conditional logic so that the user with the permission can access and submit the time entry from using others' dashboard.

## How to test (please watch the video):
1. check into current branch
2. do `npm install` and `npm run start:local` to run this PR locally
3. Clear site data/cache
4. log in with an owner account
5. go to Other Links -> Permissions Management -> Manage User Permissions -> Choose a USER without **Add Time Entry (Others)** permission under Timelog Management -> click on Add  -> Scroll down then submit
6. verify that the user who was given the permission is now able to see time entry button on another user's dashboard and submit it.

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/118868270/0c3bbd79-4b40-421b-9e39-ea7502bb9f06

## Note:
**Unmute the video**

### CODE REVIEW

**Review by pshereen — DISMISSED** [review]
Tested the PR. and the features are working as per the requirement.
Logged in as owner and added Time Entry (Others) permission to a volunteer user.

Screenshot 2024-01-23 180803

And that volunteer user was able to add a time entry form on any user via their dashboard.

Screenshot 2024-01-23 180950

Screenshot 2024-01-23 181313

**Review by Renanluizssx — CHANGES_REQUESTED** [review]
Hi @Alforoan, i followed your steps and  i logged in as volunteer, but i didn't get to see the members in the Leaderboard, i don't know what wrong i'm doing, see my review:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/91035340/7f54c413-7980-43e4-9b8b-abbbbbddbc1d

**Review by X1aoZhang99 — DISMISSED** [review]
Hi Jay, 

Just test the PR and all functionalities worded as intended. I logged in as owner and granted permission to a volunteer. The volunteer is able to add time entry for other people. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/113070229/bb3f059e-1b8d-42ea-b26c-bf4fde3d1b66

File: src/components/Timelog/Timelog.jsx [hunk 1]
```diff
@@ -530,7 +530,7 @@ const Timelog = props => {
                           </div>
                         </div>
                       ) : (
-                        canPutUserProfileImportantInfo && (
+                        (canPutUserProfileImportantInfo || canAddTimeEntry) && (
```
Comment (nathanah) [inline src/components/Timelog/Timelog.jsx:533]: Should `canAddTimeEntry` be replacing `canPutUserProfileImportantInfo` entirely or just added as it is now?

File: src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx [hunk 2]
```diff
@@ -119,7 +119,7 @@ const TimeEntryForm = props => {
   const [projectsAndTasksOptions, setProjectsAndTasksOptions] = useState([]);
   const [submitting, setSubmitting] = useState(false);
 
-  const canEditTimeEntry = props.hasPermission('editTimelogInfo') || props.hasPermission('editTimeEntry');
+  const canEditTimeEntry = props.hasPermission('editTimelogInfo') || props.hasPermission('editTimeEntry') || props.hasPermission('postTimeEntry');
```
Comment (nathanah) [inline src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx:122]: Do these need to be separated?

**Review by nathanah — CHANGES_REQUESTED** [review]
Needs a backend PR.

File: src/components/Timelog/Timelog.jsx (hunk 1 above)
Comment (Alforoan) [reply src/components/Timelog/Timelog.jsx:533]: We can just replace it entirely.

File: src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx (hunk 2 above)
Comment (Alforoan) [reply src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx:122]: Yes, I think it makes sense to separate them.

**Review by Nahiyan-16 — DISMISSED** [review]
## Works as intended
Hi Jay, I have tested your PR and it seems like all features are working as intended. I first logged in as an owner to update the add time entry (others) permissions for an admin account. After submitting, I logged into the admin account and I was able to add time for others.
Before given permissions: 
Screenshot 2024-02-01 153537
After:
[after permission given video](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/71025942/eb81e18d-0cf0-4362-b57e-1f22503a9e8e)

**Review by yvonnep165 — DISMISSED** [review]
Hi @Alforoan ,

I have tested the PR with backend PR 727 and the features are working as intended.
I logged in as owner and added Time Entry (Others) permission to a admin user. And the admin user can add time entry for other people in the team. Nice Work!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/103622026/a63dca0a-5b07-4252-a61a-2a189d3cfb31

**Review by Sophie-lei — DISMISSED** [review]
I just tested this PR and everything works as expected. after you enter the permission for time entry, the owner role can add the time entry for that user, Please view the video for details.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/70113822/c4bb1b02-915f-41e3-862d-712b1e1d88cb

**Comment by one-community** [comment]
> Hi @Alforoan, i followed your steps and i logged in as volunteer, but i didn't get to see the members in the Leaderboard, i don't know what wrong i'm doing, see my review:
> 
>  2024-01-24-00-48-31.mov

Here's why: https://www.loom.com/share/034bfdf903cd42aa9a7d7227e27e3540?sid=d6a6ad57-742d-4ecd-b3d6-5ac75dda997a

**Review by Pandani07 — DISMISSED** [review]
Tested this PR and the backend and everything works as intended. I could add the 'Add Time Entry' permission for my volunteer account and the functionality worked. Nice job! 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44672833/283f12ea-d7b9-4662-a7f7-6978bb305e24

**Review by TapanPathak — DISMISSED** [review]
Hi, I followed your steps and logged in as Owner and gave permission to Add Time Entry (Others).

Screenshot 2024-02-10 at 12 14 50 PM

Then I logged into admin account and can successfully add time for others. 

Screenshot 2024-02-10 at 12 16 19 PM

**Comment by Coops023** [comment]
Hi Jay, 
I have reviewed this PR, i logged as a volunteer to ensure i was unable to assign time entries to others, i then logged as an owner and granted the volunteer account this permission. I then logged back in as the volunteer account and i was then able to assign time entries to others. 

Perhaps worth noting in the "how to test" section that the volunteer account needs to be in a team with others to view them in the leaderboard. Great job!

Pr-1871-permissions

PR-1871-addTimeEntryOthers

**Review by Coops023 — DISMISSED** [review]

**Review by XiaohanMeng — DISMISSED** [review]
Hi, I tested this PR with the owner and admin role, and everything worked as intended! When the admin lacked permissions, they couldn't add time entries for others. Once the owner granted them permission, they were able to add time entries for others. Please see the reference video.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/113215317/ccbf175d-4d99-4620-85ec-eb4606051984

**Review by Jingyii800 — DISMISSED** [review]
Hi, I have tested this feature, it works as described. I can add Time log entry to an volunteer, and then switch to this volunteer's dashboard to add time entry. Great Jobs!. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/112589476/2875be62-d314-48f5-b821-f516f06802c5

**Review by chrischenlixing — DISMISSED** [review]
image
image
I first logged in as an admin and couldn't edit others time entry, but I logged in as an owner and then added the permission, then now I am able to edit others.

**Review by ClemarNunes — DISMISSED** [review]
I tested your PR and worked as expected.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/134466562/f591bdfe-9362-4143-bbd7-fe26e9595a60

**Review by Gugor99 — DISMISSED** [review]
### Report
I followed the steps as described, logging as owner to allow my volunteer account to see time entry. I then logged back in as a volunteer and I was able to see the option to modify time entry for others, the modal was appearing and working as expected. Well done!
### Screenshots
Cattura1
Cattura2

**Review by MALAV1308 — APPROVED** [review]
Hello, I have tested your PR and it's working perfectly as per mentioned description. I can able to add time Log after giving the permission to the User. Good Job!

**Review by brandta-1 — APPROVED** [review]
# Review
Able to edit user permissions and enter time entries on other user dashboards. 

# Video

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/116298512/8dad2b8b-527e-4355-9717-3e3c633bba33


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/116298512/6ab1cb80-8178-47db-95ce-c65de56db24a

**Review by psharma1984 — APPROVED** [review]
Hi @Alforoan Tested the functionalities and they work as expected. Was able to see the button for the user with permissions granted.
a
b

**Comment by rajnada** [comment]
Hi, I have followed the steps and reviewed the PR. It works completely fine from my side !
First, I have given permission from the Admin account.




Here bellow you can see I am login as a Developer Account and user able to add a time log.

**Review by Tiantian-C — APPROVED** [review]
It works well! Cool!

**Review by Tiantian-C — COMMENTED** [review]
I tested it, it works well!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/122494725/fdd7e3d1-00a1-4117-9b99-d36ff3e7587f

**Comment by DiegoSalas27** [comment]
Hi Jay, your changes  work as described in the PR. Great job.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/26840798/f17d2fe5-f32f-4bbe-884b-887b8939598b

**Review by Weiyao-Li — APPROVED** [review]
Hi, I have tested this PR. Definitely works as expected. One thing worth mentioning is that the permission for this add time entry is default to me, meaning that even though I am an owner, I cannot be able to delete the permission for the volunteer.
image
image

**Review by hetvi11110 — APPROVED** [review]
I tested your PR, and everything worked as mentioned. Here is the video for reviewing my testing.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/13557067/eb3781d1-f25f-4538-8f00-700959ba1117

**Review by SRamen1999 — COMMENTED** [review]
All functionalities worked as intended. I logged in as owner and granted permission to a volunteer. The volunteer is able to add time entry for other people.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/56932664/2cb70a9a-f998-4bb7-965e-d357cd90d898

File: src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx [hunk 3]
```diff
@@ -741,7 +746,8 @@ const TimeEntryForm = props => {
                   name="isTangible"
                   checked={formValues.isTangible}
                   onChange={handleInputChange}
-                  disabled={!(canEditTimeEntry || from === 'Timer')}
+                  disabled={!(canEditTimeEntry || from === 'Timer') && !(canEditTimeEntry || canAddTimeEntry)}
+
```
Comment (nathanah) [inline src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx:750]:
```suggestion
                  disabled={!(canEditTimeEntry || from === 'Timer')}
```

File: src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx [hunk 4]
```diff
@@ -676,7 +679,9 @@ const TimeEntryForm = props => {
                     placeholder="Minutes"
                     value={formValues.minutes}
                     onChange={handleInputChange}
-                    disabled={!canChangeTime}
+                    disabled={(from === 'Timer' || !(canEditTimeEntry || canAddTimeEntry)) || !canChangeTime}
+
+
```
Comment (nathanah) [inline src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx:684]:
```suggestion
                    disabled={!canChangeTime}
```

File: src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx [hunk 5]
```diff
@@ -663,7 +665,8 @@ const TimeEntryForm = props => {
                     placeholder="Hours"
                     value={formValues.hours}
                     onChange={handleInputChange}
-                    disabled={!canChangeTime}
+                    disabled={(from === 'Timer' || !(canEditTimeEntry || canAddTimeEntry)) || !canChangeTime}
+
```
Comment (nathanah) [inline src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx:669]:
```suggestion
                    disabled={!canChangeTime}

```
This logic is redundant/opaque.
`canChangeTime` is true if `from === 'TimeLog'` (on line 128) which we know is true if we're going through the "Add Time Entry for [userName]" button from line 567 from `Timelog.js`.

File: src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx [hunk 6]
```diff
@@ -642,7 +642,9 @@ const TimeEntryForm = props => {
                 id="dateOfWork"
                 value={formValues.dateOfWork}
                 onChange={handleInputChange}
-                disabled={from === 'Timer' || !canEditTimeEntry}
+
+                disabled={from === 'Timer' || !(canEditTimeEntry || canAddTimeEntry)}
+
```
Comment (nathanah) [inline src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx:647]:
```suggestion
                disabled={from === 'Timer' || !canEditTimeEntry}
```

**Review by nathanah — CHANGES_REQUESTED** [review]

**Review by Sarthak32 — APPROVED** [review]
Hi, Nice work Jay. I tested the PR and it works as expected .Once owner granted permission to a volunteer. The volunteer is able to add time entry for other people.
Screenshot 2024-03-12 211046

Screenshot 2024-03-12 211143
Screenshot 2024-03-12 211427

**Review by iven-yao — APPROVED** [review]
followed the instructions and everything works as expected.
1. enable `Add Time Entry (Others)` to a volunteer account
2. login to the volunteer role, navigate to other account's dashboard, could see there's a button as `Add Time Entry for XXX`
3. able to add a new time entry for others
Screenshot 2024-03-19 173844
Screenshot 2024-03-19 174049
image

**Comment by MQChen211** [comment]
I tested your PR; it functions as intended. I initially logged in using an Owner account and successfully assigned permissions to a Volunteer account to add time entries. I accessed the system with the Volunteer account and was able to add time entries for another account without issue. However, I did encounter a minor problem—only a select number of names are displaying on the leaderboard. Despite clearing the site's data and cache, this issue persists.

截屏2024-04-05 12 11 17

截屏2024-04-05 12 22 15

截屏2024-04-05 12 32 26

截屏2024-04-05 12 33 55

**Review by MQChen211 — APPROVED** [review]

**Pull request closed without merging**